crates/outbound-networking/src/lib.rs

```diff
@@ -289,45 +326,104 @@ fn well_known_port(scheme: &str) -> Option<u16> {
     }
 }
 
-#[derive(PartialEq, Eq, Debug, Clone)]
+#[derive(PartialEq, Debug, Clone)]
 pub enum AllowedHostsConfig {
     All,
-    SpecificHosts(Vec<AllowedHostConfig>),
+    SpecificHosts(AllowedHostsList),
+}
+
+#[derive(Debug, Clone)]
+pub enum AllowedHostsList {
+    LiteralsOnly(Vec<AllowedHostConfig>),
+    Templated(
+        Vec<TemplateableAllowedHost>,
+        Arc<spin_expressions::Resolver>,
+    ),
+}
+
+impl PartialEq for AllowedHostsList {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::LiteralsOnly(l0), Self::LiteralsOnly(r0)) => l0 == r0,
+            (Self::Templated(l0, _), Self::Templated(r0, _)) => l0 == r0,
+            _ => false,
+        }
+    }
 }
 
 impl AllowedHostsConfig {
-    pub fn parse<S: AsRef<str>>(hosts: &[S]) -> anyhow::Result<AllowedHostsConfig> {
+    pub fn parse<S: AsRef<str>>(
+        hosts: &[S],
+        resolver: &Arc<spin_expressions::Resolver>,
+    ) -> anyhow::Result<AllowedHostsConfig> {
         if hosts.len() == 1 && hosts[0].as_ref() == "insecure:allow-all" {
             bail!("'insecure:allow-all' is not allowed - use '*://*:*' instead if you really want to allow all outbound traffic'")
         }
         let mut allowed = Vec::with_capacity(hosts.len());
         for host in hosts {
-            allowed.push(AllowedHostConfig::parse(host.as_ref().to_owned())?)
+            allowed.push(TemplateableAllowedHost::parse(host.as_ref().to_owned())?);
         }
+        let allowed = if allowed
+            .iter()
+            .all(|h| matches!(h, TemplateableAllowedHost::Literal(_)))
+        {
+            let lit = AllowedHostsList::Templated(allowed, resolver.clone()).snapshot();
+            AllowedHostsList::LiteralsOnly(lit)
+        } else {
+            AllowedHostsList::Templated(allowed, resolver.clone())
+        };
```

Comment:
Could we get away with resolving eagerly here? Seems like it could potentially save a fair bit of complexity...

Reply:
I tried that in a previous iteration and it got a bit nasty because I was trying to make `DynamicHostComponent::update_data` async to allow for it and long story short the borrow checker got mad as a wet hen.  But now I know the secret sauce of `executor::block_on` I should definitely have another look - thanks!

Reply:
Oh blarg no you were correct; `executor::block_on` wouldn't be good here because a slow provider would block the entire tokio task thread.

Reply:
I reiterate that the entire `variables` edifice ought to be burned to the ground.

Reply:
A slow provider would block the Tokio task thread anyway, when it came to checking the hosts. But maybe that's more tolerable since you're about to do networking anyway?

Also, the `network` hook accesses this stuff during store build, so store build would be slowed anyway.

* Does that change your opinion on where to resolve?
* Does it mean that burning the variables edifice to the ground needs to be a prerequisite for this work?

Reply:
Blocking the thread is different (worse) than stopping the task from making progress. Tokio runs async tasks on a small pool of threads. If you block one of those threads you are taking away a sizable chunk of tokio's capacity to do concurrent work.

Reply:
Okay so between this and the newly added comment at https://github.com/fermyon/spin/pull/2299#discussion_r1513572761 it sounds like this needs to either:

* Asyncify _all_ the things, every single touchpoint for this stuff; or
* Be put on hold while we rework variables

Is that correct?